Added the capability to add custom endpoints for storage and azure active directory endpoint

Jeremy-Boyle opened the pull request:
Why:
This feature pull request allows the user to define the storage endpoints and azure active directory endpoints. This is needed for azure gov / other regions where the directory and storage endpoints aren't public and do not fall under the default provided clouds. 

Additions:
- Go Int Tests to verify that the output is as expected
- Added new backupstoragelocation config values
  - activeDirectoryAuthorityURI
- Added new volumesnapshotlocation config values
  - activeDirectoryAuthorityURI

Jeremy-Boyle commented:
I can take a look and see if there is any other use cases that might need to be added. As far as I’m aware this would allow for the custom endpoints that velero needs. 

I can look into updating the error handling to a more appropriate place.

Jeremy-Boyle commented:
> ## [Codecov](https://app.codecov.io/gh/vmware-tanzu/velero-plugin-for-microsoft-azure/pull/195?src=pr&el=h1&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=vmware-tanzu) Report
> > Merging [#195](https://app.codecov.io/gh/vmware-tanzu/velero-plugin-for-microsoft-azure/pull/195?src=pr&el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=vmware-tanzu) ([905b220](https://github.com/vmware-tanzu/velero-plugin-for-microsoft-azure/commit/905b22076fd899431c9e622d2a493871c56b6d65)) into [main](https://app.codecov.io/gh/vmware-tanzu/velero-plugin-for-microsoft-azure/commit/0fa9ea080e6d6c8e80c08ba1ec1127df958c8dcb?el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=vmware-tanzu) ([0fa9ea0](https://github.com/vmware-tanzu/velero-plugin-for-microsoft-azure/commit/0fa9ea080e6d6c8e80c08ba1ec1127df958c8dcb)) will **increase** coverage by `1.46%`.
> > The diff coverage is `63.15%`.
> 
> ```diff
> @@            Coverage Diff             @@
> ##             main     #195      +/-   ##
> ==========================================
> + Coverage   13.60%   15.07%   +1.46%     
> ==========================================
>   Files           4        4              
>   Lines         691      710      +19     
> ==========================================
> + Hits           94      107      +13     
> - Misses        594      598       +4     
> - Partials        3        5       +2     
> ```
> 
> [Impacted Files](https://app.codecov.io/gh/vmware-tanzu/velero-plugin-for-microsoft-azure/pull/195?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=vmware-tanzu)	Coverage Δ	
> [velero-plugin-for-microsoft-azure/common.go](https://app.codecov.io/gh/vmware-tanzu/velero-plugin-for-microsoft-azure/pull/195?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=vmware-tanzu#diff-dmVsZXJvLXBsdWdpbi1mb3ItbWljcm9zb2Z0LWF6dXJlL2NvbW1vbi5nbw==)	`35.00% <63.15%> (+10.40%)`	⬆️

I could increase this if you would like?

ywk253100 commented:
@Jeremy-Boyle @shubham-pampattiwar 
We introduced a new configuration item `storageAccountURI` for users to customize the storage account endpoint in PR https://github.com/vmware-tanzu/velero-plugin-for-microsoft-azure/pull/185, seems the changes in this PR contain overlap with that.
So how about that we introduce a new item `activeDirectoryAuthorityHost` to overwrite the default one if provided?

Jeremy-Boyle commented:
> storageAccountURI

This is a much better route, I will update this PR to fall in line.

Jeremy-Boyle commented:
@ywk253100 Updated based on your recommendations.

Jeremy-Boyle commented:
@ywk253100 @shubham-pampattiwar @aramprice 

Any movement ? 

Would like to get this updated ASAP

ywk253100 commented:
@Jeremy-Boyle Will review it later

ywk253100 commented on backupstoragelocation.md:
```diff
@@ -46,19 +46,25 @@ spec:
+    #
+    # Optional. This will ensure that velero uses the provided AAD URI to authenticate to the Storage Account.
+    # Note that useAAD: should be set to "true" in order to use the provided AAD URI
+    activeDirectoryAuthorityURI: my-sa.blob.core.windows.net
```
This config should also work for doing snapshots, so please add the config item to the `VolumeSnapshotLocation` and update the code.

Jeremy-Boyle replied:
Thanks, sorry for the delay, ill get to this today

ywk253100 commented on changelogs/CHANGELOG-1.7.1.md:
```diff
@@ -0,0 +1,3 @@
+## All changes
+
+- Added support for custom storage endpoints and custom active directory authority host (#195, @Jeremy-Boyle)
```
Let's remove this file.

Jeremy-Boyle commented:
Fixed Pull, there was some weird commits that were changing too much from the readme file and some merge conflicts, squashed into one commit.

Looks good now

Jeremy-Boyle commented:
@ywk253100 @shubham-pampattiwar @aramprice

Does this PR look good now ? Would like to pull these changes in, thanks!
